### PR TITLE
sql/catalog/descs: remove allocations from hot path

### PR DESCRIPTION
**Pull request by ajwerner**
The lookup by ID path gets called constantly. This was over 1% of objects allocated in some workloads. Here's a microbenchmark:

```
name                                                                    old time/op    new time/op    delta
ResolveExistingObject/CREATE_SCHEMA_sc;CREATE_TABLE_sc.foo_()sc.foo-16    2.62µs ± 1%    2.18µs ± 1%   -16.63%  (p=0.000 n=10+8)

name                                                                    old alloc/op   new alloc/op   delta
ResolveExistingObject/CREATE_SCHEMA_sc;CREATE_TABLE_sc.foo_()sc.foo-16      150B ± 0%        4B ± 0%   -97.33%  (p=0.001 n=8+9)

name                                                                    old allocs/op  new allocs/op  delta
ResolveExistingObject/CREATE_SCHEMA_sc;CREATE_TABLE_sc.foo_()sc.foo-16      12.0 ± 0%       0.0       -100.00%  (p=0.000 n=10+10)
```

Release note: None

### CODE REVIEW

**Comment by ajwerner** [comment]
@postamar this one seems like something for you to look at.

File: pkg/sql/catalog/descs/descriptor.go [hunk 1]
```diff
@@ -533,3 +552,18 @@ func (tc *Collection) deadlineHolder(txn *kv.Txn) deadlineHolder {
 func isTemporarySchema(name string) bool {
 	return strings.HasPrefix(name, catconstants.PgTempSchemaName)
 }
+
+// validationLevelMap is used to map indexes in a slice to their required
+// validation level. This avoids any heap allocations as compared to a
+// slice containing the correspondence between an index to a validation level.
+type validationLevelMap struct {
+	m util.FastIntMap
+}
```
Comment (postamar) [inline pkg/sql/catalog/descs/descriptor.go:561]: You might want to use this in the implementation of `StoredCatalog` as well!

File: pkg/sql/catalog/descs/descriptor.go [hunk 2]
```diff
@@ -86,12 +88,34 @@ func (tc *Collection) GetImmutableDescriptorByID(
...
+// into which the retreived descriptors will be stored. That slice must be the
+// same length as the ids. This allows callers seeking to get just one
+// descriptor to avoid an allocation by using a fixed-size array.
+func (tc *Collection) getDescriptorsByIDWithSlice(
+	ctx context.Context,
+	txn *kv.Txn,
+	flags tree.CommonLookupFlags,
+	descs []catalog.Descriptor,
+	ids ...descpb.ID,
+) (err error) {
```
Comment (postamar) [inline pkg/sql/catalog/descs/descriptor.go:109]: There's no point in maintaining both `getDescriptorsByID` and `getDescriptorsByIDWithSlice` now. Get rid of the former, take the latter and rename it to the former, and inline it everywhere. If need be define a helper `getDescriptorByID` which returns a `(desc, err)` pair.

**Review by postamar — CHANGES_REQUESTED** [review]
Nice, this seems like an easy win.

**Review by ajwerner — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/88614)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @postamar)

___
*[`pkg/sql/catalog/descs/descriptor.go` line 109 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/88614#-NCfwC72-zN0oAkzyqdi-r1-109:-NDVm3MPDuMwdmnc8MwR:b3uybyf) ([raw file](https://github.com/cockroachdb/cockroach/blob/6ea236f9b856c02ea64a6afbc6cc0d9dbb1515d6/pkg/sql/catalog/descs/descriptor.go#L109)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

There's no point in maintaining both `getDescriptorsByID` and `getDescriptorsByIDWithSlice` now. Get rid of the former, take the latter and rename it to the former, and inline it everywhere. If need be define a helper `getDescriptorByID` which returns a `(desc, err)` pair.
</blockquote></details>

Honestly, having two functions here is helpful. Not returning the slice is painful because you have to allocate the slice etc. I refactored it and it felt somewhat worse. I did take your advice on the single descriptor helper. I made it a function to clear up any confusion. I think it reads fine. 
___
*[`pkg/sql/catalog/descs/descriptor.go` line 561 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/88614#-NCfwC72-zN0oAkzyqdi-r1-561:-NDVlCPL4FGRS7FFpHWN:btlbz4z) ([raw file](https://github.com/cockroachdb/cockroach/blob/6ea236f9b856c02ea64a6afbc6cc0d9dbb1515d6/pkg/sql/catalog/descs/descriptor.go#L561)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

You might want to use this in the implementation of `StoredCatalog` as well!
</blockquote></details>

I'm not that worried down there -- if we're going to issue an RPC, what's a slice allocation? We can do it when a profile asks for it.

**Review by postamar — APPROVED** [review]

**Comment by ajwerner** [comment]
TFTR!

bors r=postamar

**Pull request merged**